### PR TITLE
feat: Adds support for Push-Based logging to S3

### PR DESCRIPTION
**Pull request by maastha**
## Description

Adds resource and data source for Push-Based logging to S3

NOTE: Unit tests, Github workflow updates to run acceptance tests, examples & documentations will be done in separate PRs to keep the PRs small.

Link to any related issue(s): 
[CLOUDP-240969](https://jira.mongodb.org/browse/CLOUDP-240969)
[CLOUDP-240970](https://jira.mongodb.org/browse/CLOUDP-240970)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments
image

### CODE REVIEW

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/data_source_schema.go]: i don't remember what each one was used for, do we need both description attributes?

File: internal/service/pushbasedlogexport/model.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/model.go]: as it's not needed now outside of the package (e.g. no unit tests yet) we might want to keep it private until we really need it public.  same comment for the rest

File: internal/service/pushbasedlogexport/resource.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource.go:152]: qq: what is this Timeouts when it's created or deleted?

File: internal/service/pushbasedlogexport/resource.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource.go]: as all have constants, i'd do it also for Delay

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_migration_test.go]: Recommend to use mig.CreateAndRunTest so acc and mig tests are defined in one place, e.g.:
https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/backupcompliancepolicy/resource_backup_compliance_policy_migration_test.go#L10

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_migration_test.go]: can we use ProjectIDExecution to pass the project id? or it's a global resource to the project that can cause problems?

File: internal/service/pushbasedlogexport/resource_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_test.go:22]: similar to above about using "basicTestCase" pattern with mig.CreateAndRunTest

File: internal/service/pushbasedlogexport/resource_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_test.go:24]: would it be useful to add import test as last step in basic test?

File: internal/service/pushbasedlogexport/resource_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_test.go]: as comment above, evaluate if ProjectIDExecution can be used so project_id is passed instead of projectName and orgID

File: internal/service/pushbasedlogexport/resource_test.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/resource_test.go:152]: knit: don't know if there would be a way to make more explicit in code what the comment says, in the code, like using more explicit function names

File: internal/service/pushbasedlogexport/state_transition.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/state_transition.go:45]: knit: don't know if it's possible to have only one wait function or they're different enough to have 2

**Review by lantoli — APPROVED** [review]

File: .changelog/2169.txt
Comment (AgustinBettati) [inline .changelog/2169.txt]: For new features you simply place the name of the new resource/data source/guide. Formatting is taken care by the tool. ([reference](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/changelog-process.md#header-and-entry-values)).

File: .changelog/2169.txt
Comment (AgustinBettati) [inline .changelog/2169.txt:4]: Given this will be the main feature for 1.16.0 can we make sure to update the migration guide? All good if this is handled in a follow up PR.

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: yeah this resource is configured per project so we can't use ProjectIDExecution

File: internal/service/pushbasedlogexport/resource_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_test.go:152]: any suggestions? The comment is more about how this method relates to another one.
I just wanted make it clear in comments so we don't always have to go into analyzing the code to completely understand.
I'd prefer to keep it unless there's a better approach

File: internal/service/pushbasedlogexport/state_transition.go
Comment (maastha) [reply internal/service/pushbasedlogexport/state_transition.go:45]: They are different in terms of target and pending states

File: internal/service/pushbasedlogexport/resource_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_test.go:102]: This resource is configured per project so we can't use ProjectIDExecution

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (AgustinBettati) [inline internal/service/pushbasedlogexport/resource_schema.go]: not sure if I see value in defining the links property in our terraform resources, note this is a common property Atlas API endpoints which our resources have not been exposing ([example](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Projects/operation/createProject)).

File: internal/service/pushbasedlogexport/model.go
Comment (maastha) [reply internal/service/pushbasedlogexport/model.go]: I'll leave as is for now if it's okay, to reduce merge issues with follow up PRs, as it is it can sometimes become tricky to split these PRs.
If it makes it better, I won't merge this PR until I'm very close to merging the unit tests too

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (AgustinBettati) [inline internal/service/pushbasedlogexport/resource_schema.go:50]: Do we want to preserve this default value? Would also revise if we want to define this attribute as computed.

File: internal/service/pushbasedlogexport/resource.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource.go:152]: I've kept the default timeout to 15min.
If the user doesn't specify a timeout block in their config explicitly, the default timeout value will be used.

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt:4]: Yes I'll include in a follow up PR with examples and other documentation, that ways it'll be easier to get it all reviewed by docs team as well

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (lantoli) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: does each test take very long?  it's ok as it is, or we can use ProjectIDExecution and run in serial if it doesn't take a lot of time

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (lantoli) [reply internal/service/pushbasedlogexport/resource_schema.go]: agree, i don't think links in any resource are useful to be exposed in TF (or CFN)

File: .changelog/2169.txt
Comment (JuliaMongo) [inline .changelog/2169.txt]: push-based (use lower case)

File: .changelog/2169.txt
Comment (JuliaMongo) [inline .changelog/2169.txt]: push-based

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (JuliaMongo) [reply internal/service/pushbasedlogexport/data_source_schema.go]: I would guess that you need both. cc: @zach-carr

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (JuliaMongo) [inline internal/service/pushbasedlogexport/data_source_schema.go]: Should this be "Atlas vector search", not "vector"?
Also, please use present tense:
"S3 bucket to which the vector search process writes to store logs".

File: internal/service/pushbasedlogexport/resource.go
Comment (JuliaMongo) [inline internal/service/pushbasedlogexport/resource.go]: Is this error going to be user visible? If so, please start the sentence with a capital letter, and state:
"Error when deleting a push-based log export configuration".

Please check other errors that are user visible to have their descriptions start with a capital letter and end with a period in the end.

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (JuliaMongo) [inline internal/service/pushbasedlogexport/resource_schema.go]: Same issue as before. Should this be Atlas vector search, and present tense? See my comment before where I suggested this:
"S3 bucket to which the vector search process writes to store logs".

**Review by JuliaMongo — APPROVED** [review]
LGTM for copy review with a few suggestions. Also tagging @zach-carr who might want to see this PR. Thank you!

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_schema.go]: Yeah I had second thoughts about it as well. Right now certain resource to expose this while others do not, for example organization, ldap_verify, etc do expose it. 
I checked with APIx in [Slack](https://mongodb.slack.com/archives/C05F0K6F490/p1713437455129369) and looks like it should be okay to not use these so I will remove.

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_schema.go:50]: So in TPF if attribute is marked as default it needs to be set to Computed too otherwise TF throws error. 
As for `""` this is a requirement from the API

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt]: So should this be like below?

```
```release-note:new-resource
resource/mongodbatlas_push_based_log_export
```
```

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: each test doesn't take more than a minute or two I think

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/data_source_schema.go]: https://developer.hashicorp.com/terraform/plugin/framework/handling-data/schemas#markdowndescription 
We should use both as a best practice.
Quoting from TF documentation, _"MarkdownDescription is used to provide a markdown-formatted version of the description to tooling like [terraform-plugin-docs](https://github.com/hashicorp/terraform-plugin-docs)."_

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (lantoli) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: in that case I would explore the use of ProjectIDExecution with resource.Test instead of ParallelTest

e.g. in https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/testing-best-practices.md#shared-resources

> Run the tests in serial (resource.Test instead of resource.ParallelTest) if the tests are fast and saving resources is prefered, e.g. [TestAccConfigRSProjectAPIKey_multiple](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/66c44e62c9afe04ffe8be0dbccaec682bab830e6/internal/service/projectapikey/resource_project_api_key_test.go#L149).

File: .changelog/2169.txt
Comment (AgustinBettati) [reply .changelog/2169.txt]: Simply 
```
release-note:new-resource
mongodbatlas_push_based_log_export
```

This will then generate an entry with the form of 

- **New Resource**: `mongodbatlas_push_based_log_export` 

as defined [here](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/scripts/changelog/release-note.tmpl#L3).

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (AgustinBettati) [reply internal/service/pushbasedlogexport/resource_schema.go:50]: 👍 the default "" coming for the API does seem a little odd

Comment (maastha) [reply internal/service/pushbasedlogexport/resource_schema.go:50]: right, but they are adding a mention about this in the API spec and currently this is how the UI is set up too. The idea is "" means the root directory of S3 bucket and it is a required attribute by the API so I think it makes sense.

File: internal/service/pushbasedlogexport/resource.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource.go]: added

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: added

File: internal/service/pushbasedlogexport/resource_migration_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_migration_test.go]: updated to use ProjectIDExecution with tests in serial

File: internal/service/pushbasedlogexport/resource_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_test.go:22]: updated

File: internal/service/pushbasedlogexport/resource_test.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_test.go]: updated to use ProjectIDExecution

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt]: updated

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_schema.go]: removed links

**Review by zach-carr — APPROVED** [review]
LGTM

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt]: no longer needed

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt]: no longer needed, removed

File: internal/testutil/mig/pre_check.go
Comment (lantoli) [inline internal/testutil/mig/pre_check.go:56]: is mig PreCheckAwsEnvBasic being used or it can be deleted?

File: .changelog/2169.txt
Comment (maastha) [reply .changelog/2169.txt:4]: added in PR https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2186

File: internal/testutil/mig/pre_check.go
Comment (maastha) [reply internal/testutil/mig/pre_check.go:56]: it is being used

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/data_source_schema.go]: updated

File: internal/service/pushbasedlogexport/resource_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource_schema.go]: updated

File: internal/service/pushbasedlogexport/resource.go
Comment (maastha) [reply internal/service/pushbasedlogexport/resource.go]: updated

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (lantoli) [inline internal/service/pushbasedlogexport/data_source_schema.go:23]: not for this PR, as we have the function to return the schema, we could have a helper function that for instance fill MarkdownDescription taking from Description if not filled so we only have one description in the code but it's filled for both Description and MarkdownDescription

**Review by lantoli — APPROVED** [review]
great work! and thanks for addressing the comments

File: internal/service/pushbasedlogexport/data_source_schema.go
Comment (maastha) [reply internal/service/pushbasedlogexport/data_source_schema.go:23]: makes sense, adding @AgustinBettati for visibility. Right now I do see we have the same description in 4 places

**Pull request merged**
